### PR TITLE
[apache_spark][nodes] Add Apache Spark package with Nodes data stream

### PR DESCRIPTION
**Pull request by yug-elastic**
## What does this PR do?

- Generated the skeleton of Apache Spark integration package.
- Added 1 data stream ( Nodes )
- Added data collection logic.
- Added the ingest pipelines.
- Mapped fields according to the ECS schema and added Fields metadata in the appropriate yml files.
- Added system test cases.

## Checklist

- [x] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/master/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [x] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's `changelog.yml` file.
- [x] If I'm introducing a new feature, I have modified the Kibana version constraint in my package's `manifest.yml` file to point to the latest Elastic stack release (e.g. `^7.13.0`).

## How to test this PR locally

- Clone integrations repo.
- Install elastic-package locally.
- Start elastic stack using elastic-package.
- Move to integrations/packages/apache_spark directory.
- Run the following command to run tests. 
> elastic-package test

### CODE REVIEW

**Comment by yug-elastic** [comment]
This PR is a split of https://github.com/elastic/integrations/pull/2811 as discussed over the comment https://github.com/elastic/integrations/pull/2811#issuecomment-1082701692

File: packages/apache_spark/data_stream/nodes/elasticsearch/ingest_pipeline/default.yml [hunk 1]
```diff
@@ -0,0 +1,26 @@
+---
+description: Pipeline for parsing Apache Spark nodes metrics.
+processors:
+  - set:
+      field: ecs.version
+      value: '8.0.0'
```
Comment (yug-elastic) [inline packages/apache_spark/data_stream/nodes/elasticsearch/ingest_pipeline/default.yml:6]: This will be updated with 8.2.0 after testing this integration on 8.2.0.

File: packages/apache_spark/manifest.yml [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+version: 0.1.0
+license: basic
+description: Collect metrics from Apache Spark with Elastic Agent.
+type: integration
+categories:
+  - datastore
+  - monitoring
+release: beta
+conditions:
+  kibana.version: ^8.0.0
```
Comment (yug-elastic) [inline packages/apache_spark/manifest.yml:13]: This will be updated with 8.2.0 after testing this integration on 8.2.0.

File: packages/apache_spark/_dev/build/build.yml [hunk 3]
```diff
@@ -0,0 +1,3 @@
+dependencies:
+  ecs:
+    reference: git@8.0
```
Comment (ruflin) [inline packages/apache_spark/_dev/build/build.yml:3]: Should we bump this to 8.1?

File: packages/apache_spark/changelog.yml [hunk 4]
```diff
@@ -0,0 +1,7 @@
+# newer versions go on top
+
+- version: "0.1.0"
+  changes:
+    - description: Apache Spark integration package with Main, Worker metrics
```
Comment (ruflin) [inline packages/apache_spark/changelog.yml:5]: This does not seem correct anymore?

File: packages/apache_spark/changelog.yml [hunk 5]
```diff
@@ -0,0 +1,7 @@
...
+      type: enhancement
+      link: https://github.com/elastic/integrations/pull/2811
```
Comment (ruflin) [inline packages/apache_spark/changelog.yml:7]: Same here

File: packages/apache_spark/data_stream/nodes/elasticsearch/ingest_pipeline/default.yml (hunk 1 above)
Comment (ruflin) [inline packages/apache_spark/data_stream/nodes/elasticsearch/ingest_pipeline/default.yml:6]: Should we bump this to 8.1?

File: packages/apache_spark/data_stream/nodes/fields/base-fields.yml [hunk 6]
```diff
@@ -0,0 +1,12 @@
+- name: data_stream.dataset
```
Comment (ruflin) [inline packages/apache_spark/data_stream/nodes/fields/base-fields.yml:1]: @mtojek @joshdover I remember we had a discussion around setting the value of the constant keywrods in the past but I don't remember where we landed, meaning if we should set it in the templates in package or in Fleet.

File: packages/apache_spark/data_stream/nodes/sample_event.json [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+        "ephemeral_id": "ae911caa-6091-47da-9f79-dea19b0ef5ba",
+        "id": "01cae619-f73d-4251-a6cf-9b97db33c9cd",
+        "name": "docker-fleet-agent",
+        "type": "metricbeat",
+        "version": "8.1.0"
+    },
+    "apache_spark": {
+        "metrics": {
+            "nodes": {
+                "main": {
```
Comment (ruflin) [inline packages/apache_spark/data_stream/nodes/sample_event.json:13]: Would be nice if the sample event would also contain the `worker` object.

File: packages/apache_spark/kibana/dashboard/apache_spark-b22dc960-a06c-11ec-8d4f-4fe3367a4156.json [hunk 8]
```diff
@@ -0,0 +1,1185 @@
+{
```
Comment (ruflin) [inline packages/apache_spark/kibana/dashboard/apache_spark-b22dc960-a06c-11ec-8d4f-4fe3367a4156.json:1]: Is this just the dashboards and visualisations for nodes or for everything? Lets only have the nodes part in here.

File: packages/apache_spark/changelog.yml (hunk 4 above)
Comment (yug-elastic) [reply packages/apache_spark/changelog.yml:5]: I think it's correct, the `Nodes` data stream collects the main and worker metrics, hence I have mentioned it accordingly.
Do you think I should rephrase it?

File: packages/apache_spark/data_stream/nodes/sample_event.json (hunk 7 above)
Comment (yug-elastic) [reply packages/apache_spark/data_stream/nodes/sample_event.json:13]: This `sample_event.json` was generated using the `--generate` flag during the system tests (`elastic-package test system --data-streams nodes --generate`). It would fetch any one of the `main` or `worker` and would generate the `sample_event.json`.

File: packages/apache_spark/kibana/visualization/apache_spark-038c1de0-9ee0-11ec-adb3-438cf5ea2616.json [hunk 9]
```diff
@@ -0,0 +1,102 @@
+{
```
Comment (mtojek) [inline packages/apache_spark/kibana/visualization/apache_spark-038c1de0-9ee0-11ec-adb3-438cf5ea2616.json:1]: Do you think we can extract visualizations and dashboards to a separate PR?

File: packages/apache_spark/_dev/build/docs/README.md [hunk 10]
```diff
@@ -0,0 +1,72 @@
+# Apache Spark
+
+The Apache Spark integration collects and parses data using the Jolokia Metricbeat Module.
+
+## Compatibility
+
+This module has been tested against `Apache Spark version 3.2.0`
```
Comment (mtojek) [inline packages/apache_spark/_dev/build/docs/README.md:7]: nit: integration, not module :)

File: packages/apache_spark/changelog.yml (hunk 4 above)
Comment (mtojek) [reply packages/apache_spark/changelog.yml:5]: Maybe: Implement "nodes" data stream

File: packages/apache_spark/data_stream/nodes/fields/base-fields.yml (hunk 6 above)
Comment (mtojek) [reply packages/apache_spark/data_stream/nodes/fields/base-fields.yml:1]: Let's leave it aside from this issue. I guess we have to fix it for all integrations.

File: packages/apache_spark/data_stream/nodes/fields/fields.yml [hunk 11]
```diff
@@ -0,0 +1,61 @@
+- name: apache_spark.metrics
+  type: group
+  release: beta
```
Comment (mtojek) [inline packages/apache_spark/data_stream/nodes/fields/fields.yml:3]: The release tag is deprecated, you can drop it from there and from manifest

File: packages/apache_spark/data_stream/nodes/fields/fields.yml [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+  fields:
+    - name: nodes
+      type: group
+      fields:
+        - name: main
+          type: group
+          fields:
+            - name: apps
```
Comment (mtojek) [inline packages/apache_spark/data_stream/nodes/fields/fields.yml:11]: nit: applications

**Review by mtojek — COMMENTED** [review]
I'm leaving Kibana resources for later.

File: packages/apache_spark/data_stream/nodes/fields/base-fields.yml (hunk 6 above)
Comment (joshdover) [reply packages/apache_spark/data_stream/nodes/fields/base-fields.yml:1]: +1 on having this defined in Fleet instead of every package, but agree it should be separate from this PR.

Comment (mtojek) [reply packages/apache_spark/data_stream/nodes/fields/base-fields.yml:1]: https://github.com/elastic/package-spec/issues/310

File: packages/apache_spark/_dev/build/build.yml (hunk 3 above)
Comment (yug-elastic) [reply packages/apache_spark/_dev/build/build.yml:3]: As discussed over [this comment](https://github.com/elastic/integrations/pull/2940#discussion_r839292165), do you think we should bump it up to 8.2 instead?

Comment (ruflin) [reply packages/apache_spark/_dev/build/build.yml:3]: Lets make it 8.1. 8.2 is not release yet and you should work against a released version.

File: packages/apache_spark/data_stream/nodes/sample_event.json (hunk 7 above)
Comment (ruflin) [reply packages/apache_spark/data_stream/nodes/sample_event.json:13]: I see. @mtojek I wonder if there are options in the future to have multiple sample events ;-)

File: packages/apache_spark/data_stream/nodes/fields/fields.yml (hunk 11 above)
Comment (yug-elastic) [reply packages/apache_spark/data_stream/nodes/fields/fields.yml:3]: Sure. Will do that.

File: packages/apache_spark/data_stream/nodes/sample_event.json (hunk 7 above)
Comment (mtojek) [reply packages/apache_spark/data_stream/nodes/sample_event.json:13]: https://github.com/elastic/elastic-package/issues/773

File: packages/apache_spark/kibana/visualization/apache_spark-038c1de0-9ee0-11ec-adb3-438cf5ea2616.json (hunk 9 above)
Comment (yug-elastic) [reply packages/apache_spark/kibana/visualization/apache_spark-038c1de0-9ee0-11ec-adb3-438cf5ea2616.json:1]: Yes. We can do that.

File: packages/apache_spark/kibana/dashboard/apache_spark-b22dc960-a06c-11ec-8d4f-4fe3367a4156.json (hunk 8 above)
Comment (yug-elastic) [reply packages/apache_spark/kibana/dashboard/apache_spark-b22dc960-a06c-11ec-8d4f-4fe3367a4156.json]: Makes sense, thanks! We would be raising a follow-up PR for the same as per [this](https://github.com/elastic/integrations/pull/2939#discussion_r839517284) discussion.

**Review by mtojek — APPROVED** [review]
Ship it :)
